### PR TITLE
Mariadb support

### PR DESCRIPTION
**Pull request by rmuit**
I like this particular project because it parses lines one by one (instead of doing multi-line regexps) so its time/memory consumption does not go overboard.

However, it does not properly parse logfiles generated by MariaDB.

This PR does not solve all of the issues (see https://github.com/rmuit/MySQLSlowQueryParser/tree/mariadb for more changes); it concentrates on fixing the parse loop in Parser.

Issues I found with the following representative example:
```
mysqld, Version: 10.7.1-MariaDB-1:10.7.1+maria~focal-log (mariadb.org binary distribution). started with:
Tcp port: 0  Unix socket: /run/mysqld/mysqld.sock
Time                 Id Command    Argument
mysqld, Version: 10.7.1-MariaDB-1:10.7.1+maria~focal-log (mariadb.org binary distribution). started with:
Tcp port: 0  Unix socket: /run/mysqld/mysqld.sock
Time                 Id Command    Argument
# Time: 220907 18:55:31
# User@Host: root[root] @ localhost [::1]  Id:     8
# Query_time: 0.001519  Lock_time: 0.000214 Rows_sent: 0  Rows_examined: 0
# Rows_affected: 0  Bytes_sent: 11
SELECT SOMETHING;
# User@Host: root[root] @ localhost [::1]  Id:     8
# Query_time: 0.001519  Lock_time: 0.000214 Rows_sent: 0  Rows_examined: 0
# Rows_affected: 0  Bytes_sent: 11
SELECT SOMETHING;
```
1. If two 'session headers' immediately follow each other, the second session object does not get the first 'Server' line passed in. So its `$session->Server property` is "Tcp port: ...." and its `Port` / `Transport` properties are empty.
2. Any comments up until "`# Time: `" are discarded. Since MariaDB sometimes does not always include a "`# Time: `" line (! if the query happens in the same second as the previous one?), whole entries are discarded.
3. After the "`# Time: `" comment line, exactly two extra comment lines are parsed. If the next line is still a comment (which happens with MariaDB because it has at least one extra line for "`Rows_affected: ...`", it's seen as a new entry, so
   * The original entry does not get its query included
   * Its comment lines and query are completely discarded because see point 2.

It looks to me like all of this is just an effect of quick coding and lack of tests for the edge cases. (And likely point 1 also exists for mysql log files?) However... I can't be 100% sure that there aren't existing logfiles with e.g. extra comment lines or garbage that you actually _want to_ be ignored. So I
* introduced an extra local variable for 'parse mode' which can be "" or "mariadb" based on the contents of the first server line
* made sure to keep the current behavior unchanged, _unless_ parsing in "mariadb" mode - then points 1-3 are fixed 
...to make it easier to consider this PR for inclusion.

However, if you just want these points to be fixed always... I can do that, and include tests for them. With or without simplifying other code in the big parsing 'while' loop, too.

(Regardless whether we do this / need the 'parse mode' in this loop: for proper mariadb property parsing, i.e. the code already in my repository... we'll need a 'parse mode' to change behavior inside the Session and Entry classes.)

### CODE REVIEW

File: src/PHPFUI/MySQLSlowQuery/Parser.php [hunk 1]
```diff
@@ -138,41 +138,81 @@ private function parse() : void
...
+				// if it's not a comment line, this is assumed to be the start of the
+				// next session header (i.e. there are zero queries in this session).
+				if ($parseMode === 'mariadb' && \strlen($line = $this->getNextLine()) > 0)
+					{
+					if ('#' !== $line[0])
+						{
+						$this->inSession = true;
+						}
+					$this->pushLine($line);
+					}
```
Comment (rmuit) [inline src/PHPFUI/MySQLSlowQuery/Parser.php:170]: The location for this fix is slightly awkward... which can be fixed... by reordering much of the logic in this loop.

**Comment by phpfui** [comment]
Thanks for the PR!

Yes, it has not been ported to MariaDB, so this is good. Wondering if it might make more sense to subclass MySQL and MariaDB parsing.  I am also not sure if MySQL 5.x and 8.x logs are different.

This was a quickly implemented library for my needs, but thought others might be interested, as I could not find another one. I'll take a look and see what makes sense.

**Comment by rmuit** [comment]
Re. subclassing: ...maybe 'next PR' (for Entry and Session classes).

As for this code in Parser: ...yes it looks like the library was implemented quickly. I am not seeing anything yet that must be different between different versions, and the code can be simplified; you just need to be willing to sacrifice _theoretical complete_ backward compatibility of the current behavior. (But I don't expect problems.)

**Comment by phpfui** [comment]
Have not looked at this code in a while.  It appears all the work is in the parse method.  So anything you wan to do in there is fine by me. My big issue is to not change the public interface, but it looks pretty simple and would not need to change.  Adding fields to the Entry or Session classes is fine, as those should reflect the reality of the log file, but we should probably preserve backward compatibility if a field changes.

I am turning on the slow log for my local instances. I will try to add some more examples to the tests/logs directory. Hopefully I can get some interesting edge cases, but not specifically trying to do anything exotic.

So keep adding to this PR till you are satisfied with it, then I can release.

And thanks for the help.

**Comment by rmuit** [comment]
Awesome turnaround time on a PR response!

So 

1)
The initial PR was kept small, for getting a feeling about your opinion / feedback about whether you want the Parser code to be 'ultimately backward compatible / do not change any behavior for mysql parsing' vs 'neater and simpler'.

I don't care that much one way or the other, so since my initial PR did the former... I kept it this way. IMHO We can simplify the code later if needed/wanted. (Or I can do it now if you want, given I have the code in my head right now... but I'm not opinionated on it.)

I added one extra change to Parser in this PR besides the one mentioned at point 2:
two variable name changes that I initially kept back, which IMHO make things easier to read. Because AFAICT, variable names and comments were out of sync. (`$this->inSession == true` == "not in session", `$this->inSession == false` == "in session" -> changed `$this->inSession` to `$processingSessionHeader`)

2)
I've added the work I've done, plus some quick tests:
* Session: add Version property. (I thought I needed it for the "parse type". In the end, it isn't used. But I didn't delete the code because adding it seems fine.)
* Session: add `string $parseMode` parameter to constructor, because the "TCP port:" line is subtly different on mariadb (delimiter is a double space instead of ", ")
* Entry: new properties for mariadb only. (I don't have an opinion on these extra properties being there also for mysql. I'm going to guess you don't want that -- otherwise you wouldn't have explicitly defined this property list.)
  * For this, I passed the parse mode in the previously unused `$parameters` parameter.
* Entry: the mariadb "Time:" format is different. Reformat it so `$entry->Time` is the same as the original for mysql. (Again I have no opinion but it makes sense. I took/modified this from https://github.com/HeBuguM/MySQLSlowQueryParser/tree/mariadb-fix but can't literally incorporate commits from there.)
* Parser loop: if no "# Time:" for an entry is found in mariadb mode, add the time from the previous entry.

Not done:
* Different classes per mysql version. I'm a simple guy, passing the parse mode into the constructor seemed simple enough, I feel like the "if" constructs don't make the code very messy and can't look into the future for what's best. (The messier code is actually the Parser loop, and... see point 1)
* parsing "# explain:" properly (see example in the added logfile). I don't need it and want to leave it up to a future someone to decide whether this needs a two-dimensional array or whatever. (By the way: "explain: " is not in all mariadb slow logs. It just showed up in mine at some point, without any minor version change... so I suspect there's a configuration option somewhere.)

Not planning anything else. Works For Me Now.

File: src/PHPFUI/MySQLSlowQuery/Parser.php [hunk 2]
```diff
@@ -138,54 +136,110 @@ private function parse() : void
...
+		$newSessionHeader = [];
+		$parseMode = '';
+		$previousTimeLine = '';
+		$processingSessionHeader = true;
 
+		// Get line from stack (if pushed in below loop) or file. Any comment line
+		// not starting with "# Time: " gets discarded.
 		while (\strlen($line = $this->getNextLine()))
 			{
-			if (0 === \stripos($line, self::PORT))	// in middle of session, end it
```
Comment (rmuit) [inline src/PHPFUI/MySQLSlowQuery/Parser.php:145]: Obsolete comment: "in middle of session, end it" is actually line 224-235. (That is: has always been there, since the initial repository commit.)

**Comment by phpfui** [comment]
I squashed and merged it. I have some other OS stuff to do tonight, so will take a look at this after those things.  I hope to have a new release by tomorrow.

Thanks for the PR. And yes, I wish all OS maintainers responded to things quickly. I have the time, so not a problem.
